include/nbl/ext/Scan/Scan.h

```diff
@@ -0,0 +1,190 @@
+#ifndef _NBL_EXT_SCAN_INCLUDED_
+
+#include "nabla.h"
+
+namespace nbl
+{
+namespace ext
+{
+namespace Scan
+{
+
+typedef uint32_t uint;
+#include "nbl/builtin/glsl/ext/Scan/parameters_struct.glsl"
+
+template <typename T>
+class Scan final : public core::IReferenceCounted
+{
+public:
+	_NBL_STATIC_INLINE_CONSTEXPR uint32_t DEFAULT_WORKGROUP_SIZE = 256u;
+
+	enum Operator : uint32_t
+	{
+		AND = 1 << 0,
+		XOR = 1 << 1,
+		OR  = 1 << 2,
+		ADD = 1 << 3,
+		MUL = 1 << 4,
+		MIN = 1 << 5,
+		MAX = 1 << 6,
+	};
+
+	typedef nbl_glsl_ext_Scan_Parameters_t Parameters_t;
+
+	struct DispatchInfo_t
+	{
+		uint32_t upsweep_pass_count;
+		uint32_t downsweep_pass_count;
+		uint32_t wg_count;
+		std::stack<uint32_t> element_count_pass_stack;
+	};
+
+	Scan(video::IDriver* driver, Operator op, const uint32_t wg_size) : m_wg_size(wg_size)
+	{
+		assert(nbl::core::isPoT(wg_size));
+
+		const asset::SPushConstantRange pc_range = { asset::ISpecializedShader::ESS_COMPUTE, 0u, sizeof(nbl_glsl_ext_Scan_Parameters_t) };
+		video::IGPUDescriptorSetLayout::SBinding binding = { 0u, asset::EDT_STORAGE_BUFFER, 1u, video::IGPUSpecializedShader::ESS_COMPUTE, nullptr };
+
+		m_ds_layout = driver->createGPUDescriptorSetLayout(&binding, &binding + 1);
+		m_pipeline_layout = driver->createGPUPipelineLayout(&pc_range, &pc_range + 1, core::smart_refctd_ptr(m_ds_layout));
+
+		const char* data_type_name;
+		if constexpr (std::is_same_v<T, uint32_t>)
+			data_type_name = "uint";
+		else if constexpr (std::is_same_v<T, int>)
+			data_type_name = "int";
+		else if constexpr (std::is_same_v<T, float>)
+			data_type_name = "float";
+
+		m_upsweep_pipeline = createPipeline("nbl/builtin/glsl/ext/Scan/default_upsweep.comp", op, data_type_name, driver);
+		m_downsweep_pipeline = createPipeline("nbl/builtin/glsl/ext/Scan/default_downsweep.comp", op, data_type_name, driver);
+	}
+
+	inline auto getDefaultDescriptorSetLayout() const { return m_ds_layout.get(); }
+
+	inline auto getDefaultPipelineLayout() const { return m_pipeline_layout.get(); }
+
+	inline auto getDefaultUpsweepPipeline() const { return m_upsweep_pipeline.get(); }
+
+	inline auto getDefaultDownsweepPipeline() const { return m_downsweep_pipeline.get(); }
+
+	static inline void dispatchHelper(const video::IGPUPipelineLayout* pipeline_layout, const nbl_glsl_ext_Scan_Parameters_t& params,
+		const DispatchInfo_t& dispatch_info, video::IVideoDriver* driver, bool issue_default_barrier = true)
+	{
+		driver->pushConstants(pipeline_layout, asset::ISpecializedShader::ESS_COMPUTE, 0u, sizeof(Parameters_t), &params);
+		driver->dispatch(dispatch_info.wg_count, 1, 1);
+
+		if (issue_default_barrier)
+			defaultBarrier();
+	}
+
+	static inline void buildParameters(const uint32_t in_count, Parameters_t* push_constants, DispatchInfo_t* dispatch_info,
+		const uint32_t wg_size)
```

Comment:
this function should return how many upsweeps and downsweeps to do but not in `dispatch_info`

I expect there to be one `dispatch_info` and `push_constants` value pair written out for every dispatch

Reply:
am i allowed to allocate memory here? because the caller doesn't know how many total passes would there be to do something like:
```
ScanClass::Parameters_t push_constants[total_passes];
ScanClass::DispatchInfo_t dispatch_info[total_passes];
buildParameters(.., push_constants, dispatch_info, ..);
```
same question for `RadixSort::buildParameters`?

Reply:
Most well designed libraries (such as GL) have a convention that if you pass a nullptr instead of an actual pointer to an output array, the function will do a "dry run" and not actually write anything to the pointers.. but it will however tell you how many elements it would have written.

Use the same convention

Reply:
alright